File: 1-Introductory_Problems/18-Raab_Game_I/solution.hs
import Control.Monad (replicateM_)
import Data.Bool (bool)

main :: IO ()
main = do
  t <- read <$> getLine
  replicateM_ t solveGame
    where solveGame = do
            [n, a, b] <- map read . words <$> getLine :: IO [Int]
            bool (gameOK n a b) (putStrLn "NO")
              (a + b > n || a + b > 0 && max a b >= a + b)
              where gameOK n a b = 
                      let draw = [1..(n - (a + b))]
                          aList = take (a + b) $ reverse [1..n]
                          bList = take (a + b) $ drop b $ cycle aList in
                      do
                            putStrLn "YES"
                            putStrLn $ unwords $ map show $ aList ++ draw
                            putStrLn $ unwords $ map show $ bList ++ draw
                  

File: 1-Introductory_Problems/18-Raab_Game_I/solution.cpp
#include <iostream>

int	main(void)
{
	int	t;
	int	n;
	int	a;
	int	b;

	std::cin >> t;
	while (t--)
	{
		std::cin >> n >> a >> b;
		if (a + b > n || (std::max(a, b) >= a + b && std::max(a, b) > 0))
			std::cout << "NO" << "\n";
		else
		{
			std::cout << "YES" << "\n";
			for (int i = 0; i < a + b; ++i)
				std::cout << n - i << " ";
			for (int i = 1; i <= n - (a + b); ++i)
				std::cout << i << " ";
			std::cout << "\n";
			for (int i = 0; i < a + b; ++i)
				std::cout << n - (i + b) % (a + b) << " ";
			for (int i = 1; i <= n - (a + b); ++i)
				std::cout << i << " ";
			std::cout << "\n";
		}
	}
	return (0);
}
